File: src/backend/facilitymanagement/src/main/java/org/taonaw/facilitymanagement/domain/model/equipment/Equipment.java
package org.taonaw.facilitymanagement.domain.model.equipment;

import lombok.NonNull;
import org.taonaw.facilitymanagement.domain.model.equipmentcategory.EquipmentCategoryId;

import java.util.Objects;

public class Equipment {
    private EquipmentId equipmentId;
    private EquipmentName name;
    private EquipmentCategoryId categoryId;
    private EquipmentStocks stocks;

    private Equipment() { }

    public static Equipment newEquipment(@NonNull EquipmentName equipmentName,
                                         @NonNull EquipmentCategoryId equipmentCategoryId,
                                         @NonNull EquipmentStocks equipmentStocks) {
        var equipment = new Equipment();
        equipment.equipmentId = EquipmentId.newId();
        equipment.name = equipmentName;
        equipment.categoryId = equipmentCategoryId;
        equipment.stocks = equipmentStocks;
        return equipment;
    }

    public static Equipment reconstruct(@NonNull EquipmentId equipmentId,
                                        @NonNull EquipmentName equipmentName,
                                        @NonNull EquipmentCategoryId equipmentCategoryId,
                                        @NonNull EquipmentStocks equipmentStocks) {
        var equipment = new Equipment();
        equipment.equipmentId = equipmentId;
        equipment.name = equipmentName;
        equipment.categoryId = equipmentCategoryId;
        equipment.stocks = equipmentStocks;
        return equipment;
    }

    public EquipmentId getEquipmentId() {
        return equipmentId;
    }

    public EquipmentName getName() {
        return name;
    }

    public EquipmentCategoryId getCategoryId() {
        return categoryId;
    }

    public EquipmentStocks getStocks() {
        return stocks;
    }

    public void changeName(@NonNull EquipmentName name) {
        this.name = name;
    }

    public void changeStocks(@NonNull EquipmentStocks stocks) {
        this.stocks = stocks;
    }

    public void changeEquipmentCategory(@NonNull EquipmentCategoryId categoryId) {
        this.categoryId = categoryId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Equipment equipment = (Equipment) o;
        return equipmentId.equals(equipment.equipmentId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(equipmentId);
    }
}

File: src/backend/facility/src/main/kotlin/org/taonaw/facility/domain/model/equipment/Equipment.kt
package org.taonaw.facility.domain.model.equipment

import org.taonaw.facility.domain.model.equipmentCategory.EquipmentCategoryId

class Equipment private constructor(
        val equipmentId: EquipmentId,
        val equipmentName: EquipmentName,
        val equipmentCategoryId: EquipmentCategoryId,
        val equipmentRentalFee: EquipmentRentalFee,
        val equipmentStocks: EquipmentStocks) {

    companion object {
        fun create(
                equipmentName: EquipmentName,
                equipmentCategoryId: EquipmentCategoryId,
                equipmentRentalFee: EquipmentRentalFee,
                equipmentStocks: EquipmentStocks): Equipment {

            return Equipment(EquipmentId.newId(), equipmentName, equipmentCategoryId, equipmentRentalFee, equipmentStocks)
        }
    }

    fun change(
            equipmentName: EquipmentName,
            equipmentCategoryId: EquipmentCategoryId,
            equipmentRentalFee: EquipmentRentalFee,
            equipmentStocks: EquipmentStocks): Equipment {

        return Equipment(equipmentId, equipmentName, equipmentCategoryId, equipmentRentalFee, equipmentStocks)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Equipment

        if (equipmentId != other.equipmentId) return false

        return true
    }

    override fun hashCode(): Int {
        return equipmentId.hashCode()
    }
}